11_structures/main.go

```diff
@@ -69,41 +73,323 @@ instances inside the function.
 
 E.g.
 
-    type square struct { ... }
+	type square struct { ... }
 
-    func (s square) area() float64 { ... }
-
-    func main() {
-        s := square{side: 10}
-        fmt.Println(s.area()) // 100
-    }
+	func (s square) area() float64 { ... }
 
+	func main() {
+	    s := square{side: 10}
+	    fmt.Println(s.area()) // 100
+	}
 */
 
+// TODO: Create a struct named `triangle` with fields `side1`, `side2` and
+// `side3` of type `float64`.
+type triangle struct {
+	side1 float64
+	side2 float64
+	side3 float64
+}
+
+// TODO: Create a function on the structure `area` that returns a `float64`.
+type area struct {
+	side float64
+}
+
+func (a area) areaSize() float64 {
+	return a.side * a.side
+}
+
+// TODO: Adapt your code from 02_boolean to check the triangle type. The
+// function should return a string with the type of the triangle.
+func (t triangle) triangleType() string {
+	if t.sumLengthSmallerLongestSide() {
+		return "invalid triangle values"
+	}
+	// any side <=0 return error
+	if t.side1 <= 0 || t.side2 <= 0 || t.side3 <= 0 {
+		return "invalid triangle values"
+	}
+	if t.side1 == t.side2 && t.side1 == t.side3 {
+		return "equilateral"
+	}
+	if t.side1 == t.side2 || t.side1 == t.side3 ||
+		t.side2 == t.side3 {
+		return "isosceles"
+	}
+	if t.side1 != t.side2 || t.side1 != t.side3 ||
+		t.side2 != t.side3 {
+		return "scalene"
+	}
+	return "Error 404: Triangle logic not yet found"
+}
+
+func (t triangle) longestSide() float64 {
+	if t.side1 < t.side2 && t.side2 < t.side3 {
+		return t.side3
+	}
+	if t.side1 < t.side3 && t.side3 < t.side2 {
+		return t.side2
+	}
+	return t.side1
+}
+
+func (t triangle) sumTwoShortestSides() float64 {
+	if t.side1 < t.side2 && t.side2 < t.side3 {
+		return t.side1 + t.side2
+	}
+	if t.side1 < t.side3 && t.side3 < t.side2 {
+		return t.side1 + t.side3
+	}
+	return t.side2 + t.side3
+}
+
+func (t triangle) sumLengthSmallerLongestSide() bool {
+	sum := t.sumTwoShortestSides()
+	biggest := t.longestSide()
+	return sum <= biggest
+}
+
+// TODO: Create a function `compare` on the structure `triangle` that takes another
+// triangle as an argument and returns:
+// - -1 if the area of the first triangle is smaller than the area of the second
+// - 0 if the area of the first triangle is equal to the area of the second
+// - 1 if the area of the first triangle is greater than the area of the second
+// Hint: func (t1 triangle) compare(t2 triangle) int { ... }
+
+func (t triangle) semiParameter() float64 {
```

Comment:
If I trust Wikipedia, it's actually called [Semiperimeter](https://en.wikipedia.org/wiki/Semiperimeter) :)

Reply:
I held off on renaming it semiPerimeterParameter, not sure if that would've been correct technically. 
Changed!

Reply:
I would also just call that function `semiPerimeter`. I don't see what `Parameter` would add here.